typescript/core/src/ChunkCursor.ts

```diff
@@ -211,80 +160,75 @@ export class ChunkCursor {
-        }
-      } else {
-        if (this.#startTime != undefined) {
-          startIndex = sortedIndexBy(result.record.records, this.#startTime, (logTime) => logTime);
-        }
-      }
-
-      if (startIndex >= result.record.records.length) {
-        continue;
-      }
-      if (reverse) {
-        if (
-          this.#startTime != undefined &&
-          result.record.records[startIndex]![0] < this.#startTime
-        ) {
-          continue;
-        }
-      } else {
-        if (this.#endTime != undefined && result.record.records[startIndex]![0] > this.#endTime) {
-          continue;
-        }
-      }
-
-      this.#messageIndexCursors.push({
-        index: startIndex,
-        channelId: result.record.channelId,
-        records: result.record.records,
-      });
+      this.#orderedMessageOffset = this.#orderedMessageOffset.concat(result.record.records);
     }
 
     if (offset !== messageIndexesView.byteLength) {
       throw new Error(
         `${messageIndexesView.byteLength - offset} bytes remaining in message index section`,
       );
     }
-  }
 
-  #getSortTime(): bigint {
-    if (!this.#messageIndexCursors) {
-      return this.#reverse ? this.chunkIndex.messageEndTime : this.chunkIndex.messageStartTime;
+    if (this.#orderedMessageOffset.length === 0) {
+      return;
+    }
+
+    this.#orderedMessageOffset.sort(([logTimeA], [logTimeB]) => Number(logTimeA - logTimeB));
+    if (reverse) {
+      // If we used `logTimeB - logTimeA` as the comparator for reverse iteration, messages with
+      // the same timestamp would not be in reverse order. To avoid this problem we use reverse()
+      // instead.
+      this.#orderedMessageOffset.reverse();
+    }
+
+    const [logTimeFirstMessage] = this.#orderedMessageOffset[0]!;
+    if (logTimeFirstMessage < this.chunkIndex.messageStartTime) {
+      throw new Error(
+        `Chunk contains a message with logTime (${logTimeFirstMessage}) earlier than chunk messageStartTime (${this.chunkIndex.messageStartTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+      );
     }
 
-    const cursor = this.#messageIndexCursors.peek();
-    if (!cursor) {
+    const [logTimeLastMessage] = this.#orderedMessageOffset[this.#orderedMessageOffset.length - 1]!;
+    if (logTimeLastMessage > this.chunkIndex.messageEndTime) {
       throw new Error(
-        `Unexpected empty cursor for chunk at offset ${this.chunkIndex.chunkStartOffset}`,
+        `Chunk contains a message with logTime with logTime (${logTimeLastMessage}) later than chunk messageEndTime (${this.chunkIndex.messageEndTime}) in chunk at offset ${this.chunkIndex.chunkStartOffset}`,
       );
     }
 
-    return cursor.records[cursor.index]![0];
+    // Determine the indexes corresponding to the start and end time.
+    const startTime = reverse ? this.#endTime : this.#startTime;
+    const endTime = reverse ? this.#startTime : this.#endTime;
```

Comment:
Did you try a version where you filter out entries outside of [startTime,endTime] before putting them into the array and sorting? That would certainly reduce memory usage a little bit, and reduce CPU since not as many items are being sorted – maybe not significant enough though? It would also mean you can do less bookkeeping around the endIndex.

Reply:
Before https://github.com/foxglove/mcap/pull/1037/commits/ef37a836761fc223faaf965462e9e6a03c4f6a29, I was filtering out entries that are not within the desired time range. Maybe I should go back to that and do as you say and already filter out elements of each message index before adding them to the final array.